### ISSUE TITLE
Implement Ad Preloading - Postponed

### ISSUE THREAD

**Issue by ammarahm-ed** [ammarahm-ed/react-native-admob-native-ads#82]
Initially when I made this library I worked on ad-preloading for sometime but left it off due to other bugs and rendering issues. But it seems that without ad-preloading the library is not as useful as it should be. A simple example is user staying on a screen for 3 to 4 seconds, If you want to show that ad, it takes 2 to 3 seconds to show and sometimes it is missed which means less impressions and a heap of ad-requests in AdMob. We must store the ads for at least 5 to 10 minutes on Native Code to reduce the number of requests being sent.

- This feature has been postponed.

**Comment by aminghorbankhani** [comment]
Dear Ammar Ahmed,

Thanks for your time and efforts in implementing such a great package. As we absolutely need this feature, I would appreciate it if you could give us an estimate of the time this task would take. Furthermore, is there any way to cooperate in order to get the work get done faster?

Thanks again for your consideration.

**Comment by ammarahm-ed** [comment]
I haven't got the time to work on it recently but I will very soon.

**Comment by ha-younes72** [comment]
> I haven't got the time to work on it recently but I will very soon.

That's great to hear and thanks for your time, looking forward to the new release. Would you mind pushing the native android part in a branch that we can use it before the whole other parts are done, please?

**Comment by ammarahm-ed** [comment]
@ha-younes72 you will have to wait until release since there are many undocumented breaking changes in the next version.

**Comment by ajoykarmakar** [comment]
How is it progressing?

**Comment by bartzy** [comment]
@ammarahm-ed As I understand, your plan is to prefetch some ads (5?) and store them for 5-10 minutes. When an ad is needed for display, you will just pick one ad from that cached list.
Will you remove the ad after picking it? Otherwise, it seems to me that it could be problematic from AdMob's side to show the same ad twice?

Thanks!

**Comment by ha-younes72** [comment]
@ammarahm-ed I have forked the project, and there is an Android implementation of the preloading. These are the details if you would mind: ([you can find the project here](https://github.com/ha-younes72/react-native-admob-native-ads/tree/ad-preload))

Here is a list of requirements we had in mind when working on this branch. It will be nice if you consider them as well:
1. We have a configuration number of preloaded ads, let's call it N. At the start of the app we try to load N ads. Then after each real request to show the ad, we use one of the ads in our list and try to load only 1 ad to replace the shown ad.
2.  We should try our best to show each ad only once, and reuse ads only in case of load failures. When a new ad is loaded, we have to remove one old ad from our list of ads.
3. We need Facebook Network support. For Facebook, currently, it isn't possible to load multiple ads at once, so we have to load ads in a for-loop.
4. Prevent memory leak. We have to take reference count on the number of times a UnifiedNativeAd is shown in react native components, and call its destroy method when it's reference count is zero and newer ads are in the list of pre-loaded ads.

We let the users configure multiple ad repositories so that they can have multiple lists of pre-loaded ads with different configurations including the following:
1. ad unit id: multiple-unit ids should be supported.
2. number of ads to preload.
3. should video be muted if this is a video-enabled unit.
4. should personalization be enabled.
5. Does this ad unit support Mediation, in which case the loading should be done in a for-loop.

**Comment by ha-younes72** [comment]
Dear @ammarahm-ed,

I have been testing the implemented version of the preloading branch of yours for android. but when the preloaded ad has media view, it just shows a black screen for the first preloaded ad and it shows the media view after refreshing the ad.

I will appreciate it if you have any suggestions.

**Comment by ammarahm-ed** [comment]
hey @ha-younes72 i will look into this. Can you reproduce the issue on the example branch too?

And on your previous comment, while I have implemented some part on android, the current implementation does not destroy the ads but keeps them for 5 to 10 minutes, and in that duration, those ads are shown repeatedly, so no new requests to admob server. 

I need to fix that and make sure that once an ad is shown, it should be destroyed once the view unmounts and a new ad is loaded in its place again. 
Currently there are two ArrayList of ads(one for video ads and one for image ads) and there is  configuration to specify what type of ads and how many the user wants to load on app startup. 

It needs more time and investigation to implement this feature properly, let's see.

**Comment by pairohit305** [comment]
Postponed ?? When we can expect to b completed ?

**Comment by ammarahm-ed** [comment]
@pairohit305 currently do not expect it to be completed anytime soon. Probably i will look into it again after 0.4.0 release

**Comment by pairohit305** [comment]
I tried 0.4.0 and its awesome 👏 but sadly I can’t use in real app because it take 3,4 sec to load an ad preloading is a must as ads are not preferred by user I am currently learning java so that i can create my own native packages .... I was visiting official google admob doc and they have example of preloading an ad what challenges are you facing ? to implement preloading ?

**Comment by ammarahm-ed** [comment]
time

**Comment by Hassan709** [comment]
is any one work on App Open Ads

**Comment by Vrakoss** [comment]
I would like to start to work on this feature on my free time to support a project of a friend. Where should I start though? Is it sufficient to fork https://github.com/ha-younes72/react-native-admob-native-ads/tree/ad-preload ?
